text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,371 @@
+2. Identity Providers are in control of the ID token and its contents. Unexpected changes to these contents could lead to authentication issues to crates.io.
+    - Exchanging the ID token for a crates.io token ensures we have consistency in the authorization logic of crates.io APIs.
+3. ID tokens would not benefit as much from Secret Scanning or revocation. They are JWTs with a varying format not as well-suited as the crates.io API token.
+    - By exchanging the ID token for a crates.io-specific access token, when the access token is leaked it can be quickly detected and revoked by the [existing Secret Scanning integrations](https://github.blog/changelog/2023-01-19-the-crate-io-registry-is-now-a-github-secret-scanning-integrator/).
+
+The OpenSSF Security Software Repositories Working Group [provides more detailed justification](https://repos.openssf.org/trusted-publishers-for-all-package-repositories#oidc-id-tokens-and-repository-controlled-tokens) to use the token exchange model were a valid ID token is exchanged for a package repository specific access token. 
+
+
+### What is the impact of not doing this?
+[impact-of-not-doing]: #impact-of-not-doing
+
+Foregoing the OIDC integration with trusted publishing providers would result in the status-quo. Authentication to crates.io from CI/CD systems would be facilitated via the current API token implementation.
+
+Crate authors who are more security-sensitive could using the expiring option on API tokens to further reduce potential impact of a leaked API token. However, these API tokens would need to be regularly regenerated in order to avoid interruption in release processes.
+
+These same API tokens are likely stored long term in a secrets manager like GitHub Actions Secrets or AWS Secrets Manager where a misconfiguration could lead to exposure of the secret.
+
+> [!NOTE] 
+> Other security challenges that would persist by not utilizing Trusted Publishing are detailed in the [Motivation](#motivation) section of this document.
+
+## Prior art
+[prior-art]: #prior-art
+
+Other package repositories have implemented similar support:
+
+- [PyPI's Trusted Publishing](https://docs.pypi.org/trusted-publishers/)
+- [RubyGem's Trusted Publishing](https://guides.rubygems.org/trusted-publishing/)
+- [Dart's Automated Publishing](https://dart.dev/tools/pub/automated-publishing)
+- [OSSF's Trusted Publishers For All Package Repositories Blog Post](https://repos.openssf.org/trusted-publishers-for-all-package-repositories)
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should crate owners be able to configure the allowed token scopes for a Trusted Publisher configuration?
+    - We could default to `publish-new` and `publish-update`, but maybe it's best to allow this to be configurable?
+- How long should an access token derived from the ID token exchange be valid for?
```

Comment:
do we know what the other registries have chosen and the reasons for their choices?

I guess an "issue" is that cargo by default builds the crate before it is sent to crates.io. Depending on how long this build process takes, the token may have already expired by the time cargo sends the crate to crates.io.

I wonder how viable it would be to provide a config option for the token lifetime.

Alternatively: cargo provides support for other authentication methods these days, e.g. to integrate with 1Password. Would it be viable to integrate the authentication flow through that? It would presumably allow us to request the token from crates.io just in time before the upload happens. Admittedly I don't know at what point cargo talks to the auth provider process. It might happen before the build process too. Might be worth exploring though.

Reply:
It seems like integrating this into Cargo's default set of auth providers also avoids the separate "auth" repository/action maintenance, which feels like more of a "new thing" in our release processes etc. It also seems like native integration in Cargo is probably better for users -- the risk of using the wrong repository (or an outdated version) is probably reduced.

Reply:
I guess it's up to the cargo team whether they would want to make this a built-in thing. It looks like they have quite a long to-do list already though 😅

I don't think the action would have to be integrated with the Rust release process though. All it needs is a repository with some git tags, which shouldn't be too hard to set up an maintain.

Reply:
For PyPI, the token is valid for 15 minutes from the time that it is minted. Our reasoning for this is that 15 minutes is generally enough time to perform any uploading needed, and anyone that needed more time could always re-request a token.

> I guess an "issue" is that cargo by default builds the crate before it is sent to crates.io. Depending on how long this build process takes, the token may have already expired by the time cargo sends the crate to crates.io.

I'm a little unfamiliar with the general workflow here, but it might not be necessary to exchange an OIDC token for an upload token prior to the build, i.e. this could happen after a build completes, but just before an upload will happen instead, so your only concern is the length of time of the upload.

Reply:
> 15 minutes is generally enough time to perform any uploading needed

if I understand correctly this kind of originates in PyPI allowing file uploads after a version was created? (vs. crates.io where you need to upload your one crate file upfront)



> this could happen after a build completes, but just before an upload will happen instead, so your only concern is the length of time of the upload.

the short version is `cargo publish` performs both the build and the upload and there isn't really a step in between that we could hook into from the outside (except maybe with the auth providers that I mentioned above). the uploads are currently limited to 30sec anyway due to hosting provider limitations.

Reply:
> if I understand correctly this kind of originates in PyPI allowing file uploads after a version was created? (vs. crates.io where you need to upload your one crate file upfront)

Yea, kind of, more like it originates from a PyPI release consisting of multiple files, which are uploaded in multiple separate requests rather than a single request, and that some of those files can be huge, which takes some time to upload.

> the short version is cargo publish performs both the build and the upload and there isn't really a step in between that we could hook into from the outside

I guess the question is whether the OIDC token exchange needs to happen externally (and prior to) `cargo publish` being invoked, or whether the token exchange is the responsibility of `cargo publish`, in which case it could happen between the build & publish steps.

For PyPI, we chose to do the exchange in our [canonical upload workflow](https://github.com/pypa/gh-action-pypi-publish) prior to invoking `twine upload`, but we have plans to make `twine upload` support handling the exchange directly: https://github.com/pypa/twine/issues/999. In either case, the build has already happened though, so we can keep this window relatively small.

Reply:
I think having `cargo` do the token exchange is a great idea. It'd allow us to limit the lifetime of the token to the minimum, but do not think we should take this on as a per-requisite for this work.

Instead, I'll move forward with provisioning a long enough token lifetime -- possibly configurable in the Trusted Publisher Configuration settings (up to a sane maximum allowed value). 

At the end of the `rust-lang/crates-io-auth-action` we can make a call to a revoke endpoint to invalidate the token as soon as the workflow is finished.

Reply:
I (personally) think for simplicity and security reasons it would be nicer to have this as a built-in cargo credential provider. I assume that should work, and would be easy to experiment with (you can make an external provider first to test it out). If it used libcurl, it would also share the network dependencies with cargo, keeping it lightweight. This could also work with other registries.

Reply:
> It would presumably allow us to request the token from crates.io just in time before the upload happens. Admittedly I don't know at what point cargo talks to the auth provider process.

IIRC, while you are required to provide a token in advance before `cargo publish` does a build [(see code here)](https://github.com/rust-lang/cargo/blob/351b421bed006071ecd7f79cc81335d879f2ef45/src/cargo/ops/registry/publish.rs?plain=1#L118-L128), the first token for reading registry it actually not important and the provider can just send back an [operation_independent](https://doc.rust-lang.org/nightly/cargo/reference/credential-provider-protocol.html?highlight=operation_independent#get-success-response) response. And then the provider implementation only need to deal with the publish operation [right before upload](https://github.com/rust-lang/cargo/blob/351b421bed006071ecd7f79cc81335d879f2ef45/src/cargo/ops/registry/publish.rs?plain=1#L191-L198).

It would look like this:


```mermaid
sequenceDiagram
    participant D as Developer
    participant Cargo
    participant COP as cargo-oidc-provider<br/>(built-in or installed from GH Action)
    participant OIDC as GH OIDC Provider
    participant C as crates.io
    
    D->>Cargo: 1. Trigger release workflow
    rect rgb(253, 253, 150)
    note right of Cargo: cargo publish
    Cargo->>+COP: 0. Read operation (not important)
    COP-->>-Cargo: Respond to Read operation
    Cargo-->>Cargo: Build artifact
    Cargo->>+COP: 1. Publish operation

    COP->>+OIDC: 2. Authenticate & request signed ID token
    OIDC-->>-COP: 3. Return signed ID token
    COP->>+C: 4. Send ID token & request access token
    C-->>C: 5. Verify validity of ID token using GHA OIDC's public keys
    C-->>-COP: 6. Return temporary access token
    COP-->>-Cargo: 7. Respond to Publish operation
    Cargo->>+C: 8. Publish artifact with access token
    C-->>-Cargo: Publish success
    end

    rect rgb(253, 253, 150)
    note right of Cargo: cargo logout<br/>(remote logout not implemented)
    Cargo->>+C: 9. Request revocation of access token
    C-->>-Cargo: Confirm revocation of access token
    
    end
```

With this workflow we could have an even shorter-lived access token, and can be implemented as a cargo credential provider as Eric mentioned.